File: Merge_Sort.py
def merge_sort(sequence):

    if len(sequence) <= 1:
        return sequence

    middle = int(len(sequence)/2)

    left = merge_sort(sequence[:middle])
    right =merge_sort(sequence[middle:])

    return merge(left, right)

def merge(left, right):
    end_List = []
    left_pointer = 0
    right_pointer = 0

    while left_pointer < len(left) and right_pointer < len(right):

        if left[left_pointer] < right[right_pointer]:

            end_List.append(left[left_pointer])
            left_pointer += 1

        else:
            end_List.append(right[right_pointer])
            right_pointer += 1


    end_List.extend(left[left_pointer:])
    end_List.extend(right[right_pointer:])

    return end_List
sequence = [3,2,3,1,3,4,5,6,3]
print (sequence)

print(merge_sort(sequence))

File: MergeSort.java

public class MergeSort {

	public static void main(String[] args) {
		int[] integers = {10,6,12, 4, 100, 2, 18,9};
		sort(integers, 0, integers.length-1);
		for(int i : integers){
			System.out.println(i);
		}
	}
	
	private static void sort(int[] array , int l, int r){
		if(l<r){
			int m = (l+r)/2;
			
			sort(array, l, m);
			sort(array, m+1, r);
			
			merge(array, l, r, m);
		}
	}
	
	private static void merge(int[] array , int l, int r, int m){
		int k = l;
		int n1 = m-l+1;
		int n2 = r-m;
		int[] larray = new int[n1];
		int[] rarray = new int[n2];
		for(int i=0 ; i<n1 ; ++i){
			larray[i] = array[l+i];
		}
		for(int i=0 ; i<n2 ; ++i){
			rarray[i] = array[m+i+1];
		}
		int i=0, j=0; 
		while(i<n1 && j<n2){
			if(larray[i] < rarray[j]){
				array[k++] = larray[i++];
			}else{
				array[k++] = rarray[j++];
			}
		}
		
		while(i<n1){
			array[k++] = larray[i++];
		}
		while(j<n2){
			array[k++] = rarray[j++];
		}
	}
}
